### ISSUE TITLE
Prometheus concourse_workers_registered not updated

### ISSUE THREAD

**Issue by cisco70** [concourse/concourse#4592]
## Bug Report

The Prometheus emitter set in the web node:
`
--prometheus-bind-ip=0.0.0.0 --prometheus-bind-port=9095
`
shows metrics about the stalled workers that do not reflect the actual status of the cluster.
This is what the fly command shows:
`
$ fly -t cd workers
name                 containers  platform  tags     team  state    version  age
i-00bef60f61f3140f1  140         linux     none     none  running  2.2      1d 
i-05188cf705b4d69c3  144         linux     none     none  running  2.2      1d 
i-0a80fcf7700e58d7e  165         linux     none     none  running  2.2      2d 
i-0babe95ca11164476  135         linux     none     none  running  2.2      2d 
i-0e9092dc0af1f7742  21          linux     frankie  none  running  2.2      2d 
i-0ee8c2d0d45117352  140         linux     none     none  running  2.2      1d 
i-0f21ac10508ef79bb  147         linux     none     none  running  2.2      2d 
i-0f2d607567b707073  159         linux     none     none  running  2.2      2d 
i-0f4323cc9761eb310  160         linux     none     none  running  2.2      1d 
`

This what the emitter reports:
 
`
curl -s localhost:9095 | grep concourse_workers_registered
...
concourse_workers_registered{state="running"} 9
concourse_workers_registered{state="stalled"} 7
`

## Steps to Reproduce

1) I've created a playground environment to reproduce it. List the workers:
`
$ fly -t cd-pre-service workers
name                 containers  platform  tags  team  state    version  age
i-0ef1327eed5260571  1           linux     none  none  running  2.2      6d 
`
2) Query the concourse metric:
`
curl -s localhost:9095/metrics | grep registered
...
concourse_workers_registered{state="running"} 1
concourse_workers_registered{state="stalled"} 0
`
3) force the stalled status stopping the concourse worker process
4) prune the stalled worker:

`
fly -t cd-pre-service prune-worker -w i-0ef1327eed5260571 
pruned 'i-0ef1327eed5260571'
`

5) repeat 1 and 2

## Expected Results

The stalled worker should not appear in the metrics

## Actual Results

`
 curl -s localhost:9095/metrics | grep registered
...
concourse_workers_registered{state="running"} 1
concourse_workers_registered{state="stalled"} 1
`
## Version Info

* Concourse version: 5.6
* Deployment type (BOSH/Docker/binary): binary
* Infrastructure/IaaS: AWS EC2

**Comment by pivotal-jamie-klassen** [comment]
Conceptually this feels similar to #4341 which should have been resolved by #4446 which made it into v5.6.0. The prometheus metrics emitter has its own cache eviction policy where it cleans up at least some of the metrics from workers that have no longer been seen after 5 minutes:

https://github.com/concourse/concourse/blob/master/atc/metric/emitter/prometheus.go#L709

I'm wondering if this metric eventually disappears after 5 minutes or if it sticks around forever? i haven't had time to test today, but I'd be curious about that. @kmdouglass could probably respond to this a bit more confidently.

In any case, if these metrics are never getting cleaned up I agree this is a bug, but if they're just taking 5 minutes to clean up it might be more of a feature request? that pruning should clear the prometheus metrics emitter.

**Comment by beoboo** [comment]
To add more details: one of our workers stalled. The metric above (that showed 7 stalled workers) was then updated

```
curl -s localhost:9095 | grep concourse_workers_registered ... concourse_workers_registered{state="running"} 8 
concourse_workers_registered{state="stalled"} 1
```

Then we pruned the worker, and the metric stayed the same. After a while, the autoscaling group spawned another worker, and the metric became:

```
curl -s localhost:9095 | grep concourse_workers_registered ... concourse_workers_registered{state="running"} 9
concourse_workers_registered{state="stalled"} 1
```

so still indicating one stalled worker, even after more than 5 minutes.

**Comment by kmdouglass** [comment]
@pivotal-jamie-klassen Indeed this looks like the same sort of problem as described in #4341 . Were the `concourse_workers_registered` metrics added around the same time that #4446 was merged? I didn't see this metric in my testing.

**Comment by kmdouglass** [comment]
I looked at the diff from #4446 . What's strange is that the `concourse_workers_registered` metric was neither garbage collected before the fix, nor did we see this metric outliving workers in RC  7c63f45 that we were running production at the time.

The snippet from the diff below shows that only the `workers_container`, `volumes`, and `tasks` metrics were originally intended to be garbage collected.

```
$ git diff 591268378 c63baf150

<snip>

 func (emitter *PrometheusEmitter) httpResponseTimeMetrics(logger lager.Logger, event metric.Event) {
@@ -649,14 +707,7 @@ func (emitter *PrometheusEmitter) periodicMetricGC() {
                now := time.Now()
                for worker, lastSeen := range emitter.workerLastSeen {
                        if now.Sub(lastSeen) > 5*time.Minute {
-                               // This is a little stupid but we don't know the platform here,
-                               // but DeleteLabelValues requires an exact match on the label set.
-                               // As a workaround we try all known values for  the "platform" label
-                               for _, platform := range []string{"linux", "windows", "darwin"} {
-                                       emitter.workerContainers.DeleteLabelValues(worker, platform)
-                                       emitter.workerVolumes.DeleteLabelValues(worker, platform)
-                                       emitter.workerTasks.DeleteLabelValues(worker, platform)
-                               }
+                               DoGarbageCollection(emitter, worker)
                                delete(emitter.workerLastSeen, worker)
                        }
                }
@@ -664,3 +715,57 @@ func (emitter *PrometheusEmitter) periodicMetricGC() {
                time.Sleep(60 * time.Second)
        }
 }
+
+// DoGarbageCollection retrieves and deletes stale metrics by their labels.
+func DoGarbageCollection(emitter PrometheusGarbageCollectable, worker string) {
+       for _, labels := range emitter.WorkerContainersLabels()[worker] {
+               emitter.WorkerContainers().Delete(labels)
+       }
+
+       for _, labels := range emitter.WorkerVolumesLabels()[worker] {
+               emitter.WorkerVolumes().Delete(labels)
+       }
+
+       for _, labels := range emitter.WorkerTasksLabels()[worker] {
+               emitter.WorkerTasks().Delete(labels)
+       }
+
+       delete(emitter.WorkerContainersLabels(), worker)
+       delete(emitter.WorkerVolumesLabels(), worker)
+       delete(emitter.WorkerTasksLabels(), worker)
+}

<snip>
```

**Comment by pivotal-jamie-klassen** [comment]
This was a dupe of #3856, and is fixed by #4895 in v5.8.0
